proposals/XXXX-ConcurrentValueAndConcurrentClosures.md

```diff
@@ -0,0 +1,720 @@
+
+
+#### Actor types
+
+Actor types provide their own internal synchronization, so they implicitly conform to `ConcurrentValue`. The [actors proposal](https://github.com/DougGregor/swift-evolution/blob/actors/proposals/nnnn-actors.md) provides more detail.
+
+
+#### Key path literals
+
+Key paths themselves conform to the `ConcurrentValue` protocol. However, to ensure that it is safe to share key paths, key path literals can only capture values of types that conform to the `ConcurrentValue` protocol. This affects uses of subscripts in key paths:
+
+
+```
+class SomeClass: Hashable {
+  var value: Int
+}
+
+class SomeContainer {
+  var dict: [SomeClass : String]
+}
+
+let sc = SomeClass(...)
+
+// error: capture of 'sc' in key path requires 'SomeClass' to conform
+// to 'ConcurrentValue'
+let keyPath = \SomeContainer.dict[sc]   
+```
+
+
+
+### New `@concurrent` attribute for functions
+
+While the `ConcurrentValue` protocol directly addresses value types and allows classes to opt-in to participation with the concurrency system, function types are also important reference types that cannot currently conform to protocols. Functions in Swift occur in several forms, including global func declarations, nested functions, accessors (getters, setters, subscripts, etc), and closures.  It is useful and important to allow functions to be passed across concurrency domains where possible to allow higher order functional programming techniques in the Swift Concurrency model, for example to allow definition of `parallelMap` and other obvious concurrency constructs.
+
+We propose defining a new attribute on function types named `@concurrent`.   A `@concurrent` function type is safe to transfer across concurrency domains (and thus, it implicitly conforms to the `ConcurrentValue` protocol).  To ensure memory safety, the compiler checks several things about values (e.g. closures and functions) that have `@concurrent` function type:
+
+
+
+1. A function can be marked `@concurrent`. Any captures must also conform to `ConcurrentValue`.
+2. Closures that have `@concurrent` function type can only use by-value captures. Captures of immutable values introduced by `let` are implicitly by-value; any other capture must be specified via a capture list: \
+`let prefix: String = ... \
+var suffix: String = …  \
+strings.parallelMap { [suffix] in prefix + $0 + suffix }` \
+The types of all captured values must conform to `ConcurrentValue`.
+3. Accessors are not currently allowed to participate with the `@concurrent` system as of this proposal.  It would be straight-forward to allow getters to do so in a future proposal if there was demand for this.
```

Comment:
```suggestion
1.  A function can be marked `@concurrent`. Any captures must also conform to `ConcurrentValue`.

2.  Closures that have `@concurrent` function type can only use by-value captures. Captures of immutable values introduced by `let` are implicitly by-value; any other capture must be specified via a capture list:

    ```swift
    let prefix: String = ...
    var suffix: String = ...
    strings.parallelMap { [suffix] in prefix + $0 + suffix }
    ```

    The types of all captured values must conform to `ConcurrentValue`.

3.  Accessors are not currently allowed to participate with the `@concurrent` system as of this proposal.  It would be straight-forward to allow getters to do so in a future proposal if there was demand for this.
```